계산기 [STEP 2] Kobe

devKobe24 opened the pull request:
안녕하세요 원비(@wonbi92) 🙌

STEP2 에서 많은 것을 배운 코비🫠 입니다 :)
참으로 많은 것을 배울수 있었던 STEP2 였습니다.
정말 재미있고 즐겁게 STEP 2를 진행했던것 같습니다.

우여곡절 끝에 드디어 STEP 2 PR을 보냅니다 🙋‍♂️

---

### 🤔 고민한 점
1. UML만을 보고 내부 구조를 생각해내는 것에 대한 고민이 많았습니다.
- UML을 처음 보고 느낀 감정은 '아무것도 설명이 안되어있는데 어떻게 구현하지?' 였습니다. 그래서 일단 TDD를 시행하였고 TDD를 시행하면서 점점 실마리가 보여가며 엉켜있던 실들을 풀어가는 느낌이였습니다.

2. TDD의 객체 구현
- 원비께서 주신 피드백 중 아에 다른 Mock 객체가 만들어졌다는 피드백을 듣고 실제 객체와 어떻게 연결해서 실험을 해야하는지 고민이 많았습니다. 제가 선택한 방법은 실제 객체 중 LinkedList 프로토콜을 Mock 객체와 연결하여 해결하였습니다.

3. 제네릭
- Operands와 Operator의 각각의 타입을 받아야하는데 어떻게 받아야 할지 생각을 못했습니다. 그래서 학습을 해본 결과 여러개의 제네릭을 받을 수 있다는 사실을 알게되었습니다. 그래서 각각의 제네릭을 나눠서 받아 해결하였습니다.

---

### 🙋‍♂️ 조언을 얻고 싶은 부분

1. UML을 보고 내부 구현을 파악하는 방법.
- 이번 스탭에서 가장 힘들었던 부분이 UML만 보고 내부 구현 코드를 만드는 것이였습니다. 그러면서 느낀점이 'UML을 보고 내부 구현을 어떻게 파악하지?' 가 가장 궁굼하고 조언을 얻고 싶은 부분이였습니다.

Commit:
:sparkles: Feat: 열거형 ExpressionParser 구현 및 파일 생성.

ExpressionParser 파일을 생성 후 빈 열거형으로 ExpressionParser를 구현을 했습니다.

Commit:
:sparkles: Feat: Formula 파일 생성및 구조체 생성.

Formula 파일 생성 후 빈 구조체를 구현하였습니다.

Commit:
:label: Add: FormulaProtocol 구현.

FormulaProtocol을 구현하였습니다.

Commit:
:poop: Add: Formula 구조체 임시 구현.

Formula 구조체를 임시적으로 구현하였습니다.

Commit:
:poop: Add: ExpressionParser 열거형 임시구현.

ExpressionParser 열거형을 임시적으로 구현하였습니다.

Commit:
:sparkles: Feat: MockFormula 객체 생성.

테스트를 위한 MockFormula 객체를 생성하였습니다.

Commit:
:recycle: Refactor: Double를 확장, CalculateItem 채택.

Double을 확장시켜 CalculateItem을 채택하게 만들었습니다.

Commit:
:poop: Add: Operator 열거형 임시 구현, Operator 확장.

Operator 열거형을 임시로 구현하였습니다, Operator 열거형을 확장하여 메서드들을 임시로 구현하였습니다. Operator 열거형을 확장하여 CalculateItem 프로토콜을  채택하였습니다.

Commit:
:sparkles: Feat: String Extension 파일 생성 및 임시 구현.

String Extension 파일을 생성하고 임시로 구현하였습니다.

Commit:
:recycle: Refactor: calculate 메서드 수정.

Operator 열거형의 calculate 메서드 내부 로직을 수정하였습니다.

Commit:
:recycle: Refactor: 제네릭 CalculateItem 채택, 불필요 코드 삭제.

제네릭 T 타입이 CalculateItem을 채택하게 만들었습니다 그로 인해 불필요해진 CalculateItemProtocol을 삭제했고, MockLinkedListProtocol도 삭제하였습니다.

Commit:
:recycle: Refactor: LinkedListProtocol 채택.

MockLinkedListProtocol을 채택했던 것을 삭제하고 LinkedListProtocol을 채택했습니다. 이로 인해 테스트의 신뢰도와 테스트 더블의 의도와 적합해졌습니다.

Commit:
:sparkles: Feat: Operator 테스트를 위한 Mock 객체 생성.

Operator 열거형 테스트를 위해 MockOperator 객체를 생성했습니다.

Commit:
:white_check_mark: Test: Operator의 add 테스트[GREEN]

Operator의 add의 calculate 호출시 두 값이 더해지는지 테스트를 했고 그 결과 값이 더해지는 것을 확인했습니다.

Commit:
:recycle: Refactor: 코드 정리, add 테스트, Mock객체 생성.

Commit:
:white_check_mark: Test: Operator subtract 호출시 테스트[GREEN]

MockOperator의 subtract 호출시 두 값이 빼기 연산이 실행되어 그 결과값을 반환하는 것을 확인했습니다.

Commit:
:white_check_mark: Test: Operator의 divide 호출시 테스트[GREEN]

MockOperator의 divide 호출시 두 값이 나누기 연산이 되고 그 결과값을 반환하는지 확인하는 테스트를 시행하였고 테스트에 성공하였습니다.

Commit:
:white_check_mark: Test: Operator의 multiply 테스트[GREEN]

MockOperator의 multiply 호출시 두 값을 곱하는 연산을 하여 결과값을 반환하는 테스트를 시행하였고 테스트에 성공하였습니다.

Commit:
:recycle: Refactor: 프로토콜, 구조체 내부 로직 수정.

MockFormulaProtocol의 내부 코드를 수정했습니다. MockFormula의 내부 코드를 수정했습니다.

Commit:
:test_tube: Test: result 호출시 operator에 따른 결과값 반환 테스트[RED]

MockFormula의 result 호출시 operator가 플러스일 경우 받아온 값을 더하기 연산하여 결과값을 반환하는 테스트를 시행하였고 실패하였습니다.

Commit:
:recycle: Refactor: 프로토콜, 구조체 내부 코드 수정, 테스트 실패

Commit:
:recycle: Refactor: String Extension 로직 변경.

String Extension의 함수 split의 내부 로직을 변경하였습니다.

Commit:
:pencil2: Fix: 불명확한 테스트 함수 이름들 수정.

불명확하게 설정된 함수 이름들을 수정하였습니다.

Commit:
:recycle: Refactor: 테스트를 위하여 프로토콜과 구조체 코드 수정.

케이스 테스트 비교를 위하여 MockFormulaProtocol의 operands와 operators의 타입을 옵셔널로 변경하였으며 그에 따라 MockFormula 구조체도 변경해주었습니다.

Commit:
:sparkles: Feat: ExpressionParser의 Mock 객체 생성.

테스트를 위형 MockExpressionParser 객체를 생성하였습니다.

Commit:
:white_check_mark: Test: operator 초기값 할당 테스트[GREEN]

MockFormula에 operators에 add를 초기화 할 수 있는지 테스트를 실행하였고 성공하였습니다.

Commit:
:art: Refactor: Mock객체 생성, add 테스트, 프로토콜과 구조체 수정.

Commit:
:white_check_mark: Test: MockFormula operator subtract 테스트[GREEN]

MockFormula operator의 초기값으로 subtract가 들어가는지 테스트를 실행했고 성공했습니다.

Commit:
:white_check_mark: Test: MockFormula operator divide 테스트[GREEN]

MockFormula operator의 초기값으로 divide가 들어가는지 테스트를 실행하였으며 성공했습니다.

Commit:
:white_check_mark: Test: MockFormula operator multiply 테스트[GREEN]

MockFormula operator의 초기값으로 multiply가 들어가는지 테스트를 실행하였으며 성공했습니다.

Commit:
:recycle: Refactor: split 내부 코드 수정.

String Extension 내부에 있는 split 함수의 코드를 수정하였습니다.

Commit:
:white_check_mark: Test: MockFormula operands 테스트[GREEN]

MockFormula operands에 초기값을 할당할 수 있는지에 대한 테스트를 시행하였으며 성공했습니다.

Commit:
:recycle: Refactor: MockFormula result 내부 코드 수정.

케이스 테스트를 위해 result 내부의 코드를 수정하였습니다.

Commit:
:white_check_mark: Test: MockFormula result operands 반환 테스트[GREEN]

MockFormula result 메서드를 호출했을 때 operands의 값이 반환되는지 테스트 하였으며 성공하였습니다.

Commit:
:recycle: Refactor: MockFormula result 메서드 내부 코드 수정.

테스트를 위해 MockFormula의 result 메서드 내부 코드를 수정하였습니다.

Commit:
:test_tube: Test: MockFormula result operands 반환 테스트[RED]

MockFormula의 result 메서드를 호출했을 경우 모든 큐에 들어간 값들이 반환되는지 확인하는 테스트를 시행하였고 실패하였습니다.

Commit:
:recycle: Refactor: MockFormula result 메서드 로직 변경.

MockFormula의 result 메서드 내부 로직을 변경하였습니다.

Commit:
:white_check_mark: Test: MockFormula results 연산자 add 테스트[GREEN]

MockFormula의 result 메서드를 호출시 연산자가 add이면 피연산자를 합한 값을 반환하는 테스트를 시행하였고 성공했습니다.

Commit:
:recycle: Refactor: test 케이스 코드 수정.

MockFormula results 호출시 operands 값이 반환되는 테스트 케이스 코드를 수정했습니다.

Commit:
:white_check_mark: Test: MockFormula results subtract 호출 테스트[GREEN]

MockFormula의 results 호출시 연산자가 subtract일 경우 각각의 피연산자를 뺀 결과값을 반환하는 테스트를 실행하였고 성공하였습니다.

Commit:
:white_check_mark: Test: MockFormula results divide 테스트[GREEN]

MockFormula의 results 호출시 연산자가 divide일 경우 두 피연산자를 나눈 결과값을 반환하는 테스트를 시행하였고 성공하였습니다.

Commit:
:white_check_mark: Test: MockFormula results multiply 테스트[GREEN]

MockFormula의 results 호출시 연산자가 multiply일 경우 두 피연산자를 곱한 결과값을 반환하는 테스트를 시행하였고 성공했습니다.

Commit:
:pencil2: Fix: 잘못된 띄어쓰기 수정.

중괄호와 소괄호 사이가 띄어져있지 않은것을 띄워주는 것으로 수정하였습니다.

Commit:
:recycle: Refactor: componentsByOperators 함수 로직 변경.

테스트를 위해 MockExpressionParser의 componentsByOperators의 함수 로직을 변경해주었습니다.

Commit:
:test_tube: Test: componentsByOperator 호출 테스트[RED]

MockExpressionParser의 componentsByOperator를 호출시 String으로 들어간 연산 표현이 연산자를 기준으로 잘라져서 반환되는 테스트를 시행했으며 실패했습니다.

Commit:
:recycle: Refactor: componentsByOperators 내부 로직 변경.

MockExpressionParser의 componentsByOperators 함수의 내부 로직을 변경하였습니다.

Commit:
:white_check_mark: Test: componentsByOperators 호출 테스트[GREEN]

MockExpressionParser의 componentsByOperators 함수 호출시 String로 들어간 연산 표현이 하나씩 나뉘어 배열에 들어가 반환되는 테스트를 시행하였고 성공하였습니다.

Commit:
:recycle: Refactor: componentsByOperators 메서드 로직 수정.

MockExpressionParser의 componentsByOperators 메서드 로직을 수정했습니다.

Commit:
:white_check_mark: Test: componentsByOperators 메서드 호출 테스트[GREEN]

MockExpressionParser의 componentsByOperators 호출시 String으로 들어온 연산 표현식이 연산자 빼고 배열에 담겨 반환되는것을 테스트 했고 성공했습니다.

Commit:
:recycle: Refactor: componentsByOperators 메서드 내부 로직 수정.

MockExpressionParser의 componentsByOperators의 메서드 내부의 로직을 고차함수와 split을 이용하여 수정하였습니다.

Commit:
:recycle: Refactor: parse 메서드 내부 로직 수정.

MockExpressionParser의 parse 메서드 내부 로직을 수정하였습니다.

Commit:
:white_check_mark: Test: parse 메서드 호출시 MockFormula 객체 반환 테스트[GREEN]

MockExpressionParser의 parse 메서드 호출시 MockFormula 객체가 반환되는지 테스트하였고 성공하였습니다.

Commit:
:sparkles: Feat: Protocols 파일생성, LinkedListProtocol 파일 생성 및 구현.

Protocol 들만 모여있게 만들기 위해 파일을 생성하였으며 LinkedListProtocol을 새로 파일 생성하여 옮겼습니다.

Commit:
:sparkles: Feat: CalculateItem 파일생성 및 구현.

Protocols 폴더에 하위로 CalculateItem 파일을 생성하고 CalculateItem 프로토콜을 코드를 옮겼습니다.

Commit:
:sparkles: Feat: Protocols 하위에 FormulaProtocol 파일 생성.

Protocols 하위에 FormulaProtocol 파일을 생성하고 프로토콜 코드를 옮겼습니다.

Commit:
:sparkles: Feat: Extensions 폴더 생성 및 String+Extension 파일 이동.

Extensions 폴더를 생성하고 그 하위에 String+Extension 파일을 이동시켰습니다.

Commit:
:sparkles: Feat: Doule+Extension 파일 생성 및 코드 이동.

Extensions 폴더 하위에 Double+Extension 파일을 생성했고 내부에 Double Extension 코드를 이동시켰습니다.

Commit:
:sparkles: Feat: Operator+Extension 파일 생성 및 코드 이동.

Extensions 폴더 하위에 Operator+Extension 파일을 생성했으며 내부로 코드를 이동 시켰습니다.

Commit:
:recycle: Refactor: Operator+Extension 코드 수정.

분리되어있던 Operator+Extension 코드를 하나로 합쳤습니다.

Commit:
:recycle: Refactor: componentsByOperators 내부 코드 수정[BLUE].

ExpressionParser의 componentsByOperators 내부 코드를 수정하였습니다.

Commit:
:fire: Refactor: 불필요한 주석 삭제.

불필요한 주석을 삭제해주었습니다.

Commit:
:recycle: Refactor: FormulaProtocol 코드 수정.

FormulaProtocol의 내부에 U 제네릭 타입을 추가하고 operators의 CalculatorItemQueue에 U를 제네릭을 채택했습니다.

Commit:
:recycle: Refactor: Formula 구조체 코드 수정.

Formula 구조체가 T와 U를 제네릭을 받는 코드로 수정하였습니다.

Commit:
:recycle: Refactor: Formula의 operators 제네릭 타입 수정[BLUE]

Formula의 operators의 CalulatorItemQueue<T> 타입에서 CalulatorItemQueue<U> 타입으로 수정하였습니다.

Commit:
:recycle: Refactor: ExpressionParser의 parse 내부 로직 수정[BLUE]

parse 메서드의 내부 로직을 테스트 성공에 맞는 조건으로 수정하였습니다.

Commit:
:recycle: Refactor: Formula의 result 함수 내부 로직 수정[BLUE].

Formula의 result 함수의 내부 로직을 테스트 성공 케이스에 부합하도록 수정하였습니다.

Commit:
:memo: Add: STEP2 계산기 README Upload.

계산기 프로젝트 STEP2의 README를 올립니다.

wonbi92 commented on Calculator/Calculator/Model/ExpressionParser.swift:
```diff
@@ -0,0 +1,42 @@
+//  Created by Minseong Kang on 2023/06/06.
+//
+
+enum ExpressionParser<T: CalculateItem, U: CalculateItem> {
```
제네릭타입 T와 U는 어디에서 사용되는 타입일까요? 👀

wonbi92 commented on Calculator/Calculator/Model/ExpressionParser.swift:
```diff
@@ -0,0 +1,42 @@
+		}
+		
+		let operatorList = Operator.allCases.compactMap { $0.rawValue }
+		operatorList.forEach {
+			guard input.contains($0) else { return }
+			guard let `operator` = Operator(rawValue: $0) else { return }
+			operatorQueue.enqueue(`operator`)
+		}
```
현재의 로직대로라면, `"1+2+3"`과 같이 같은 연산자가 2개 이상 들어올 경우에 이를 제대로 처리할 수 없을거 같아요

wonbi92 commented on Calculator/Calculator/Model/Extensions/Operator+Extension.swift:
```diff
@@ -0,0 +1,40 @@
+			}
+		}
+		return operatorsCalculateResult
+	}
```
지금의 로직대로라면, `operatorsCalculateResult`변수는 불필요해보이네요
```suggestion
    func calculate(lhs: Double, rhs: Double) -> Double {
        switch self {
        case .add:
            return add(lhs: lhs, rhs: rhs)
        case .subtract:
            return subtract(lhs: lhs, rhs: rhs)
        case .divide:
            return divide(lhs: lhs, rhs: rhs)
        case .multiply:
            return multiply(lhs: lhs, rhs: rhs)
        }
    }
	
```

wonbi92 commented on Calculator/Calculator/Model/Extensions/Operator+Extension.swift:
```diff
@@ -0,0 +1,40 @@
+		return lhs - rhs
+	}
+	
+	private func divide(lhs: Double, rhs: Double) -> Double {
+		return lhs / rhs
+	}
```
0으로 나눌경우 어떻게 될까요? 👀

wonbi92 commented on Calculator/Calculator/Model/Extensions/String+Extension.swift:
```diff
@@ -0,0 +1,16 @@
+//  Created by Minseong Kang on 2023/06/06.
+//
+
+import Foundation
```
`Foundation`을 임포트 하신 이유를 알 수 있을까요? 👀

wonbi92 commented on Calculator/Calculator/Model/Extensions/String+Extension.swift:
```diff
@@ -0,0 +1,16 @@
+	func split(with target: Character) -> [String] {
+		let splitByTarget = self.split(separator: target).map { String($0) }
+
+		return splitByTarget
```
이부분 역시 굳이 상수로 선언한 후에 반환할 필요는 없어보입니다!
```suggestion
		return self.split(separator: target).map { String($0) }
```

wonbi92 commented on Calculator/Calculator/Model/Formula.swift:
```diff
@@ -0,0 +1,40 @@
+		case .multiply:
+			return lhsOperand * rhsOperand
+		}
+	}
```
현재 `result()`메서드의 동작이 정확하게 이해가 잘 되지 않습니다..! `Formula` 타입은 입력받은 피연산자들과 연산자들을 순차적으로 계산하는 역할을 합니다! `result()`메서드는 이를 실행하는 메서드가 되겠네요! 

쉬운 예를 하나 드리겠습니다.
1. `operands`에 [1, 2, 3] 이 들어가있고, `operators`에 [+, *] 이 들어가 있으면 
2. 순차적으로 나열되어 `1 + 2 * 3`이 되고, 
3. 이는 9라는 결과값을 반환하겠네요!

wonbi92 commented on Calculator/Calculator/Model/Formula.swift:
```diff
@@ -0,0 +1,40 @@
+	func result() -> Double {
+		var operandsList: [Double] = []
+		while let operands = operands.dequeue() {
+			guard let operand = operands as? Double else { return -1.0 }
```
가드문을 통과하지 못했을 때 -1.0을 반환하는 이유가 무엇일까요? 👀

wonbi92 commented on Calculator/Calculator/Model/Formula.swift:
```diff
@@ -0,0 +1,40 @@
+		let lhsOperand = operandsList[operandsList.startIndex]
+		let rhsOperand = operandsList[operandsList.endIndex-1]
+
+		guard let operators = operators.dequeue() as? Operator else { return -3.0 }
```
마찬가지로 가드문을 통과하지 못했을 때 -3.0을 반환하는 이유가 무엇일까요? 👀

wonbi92 commented on Calculator/Calculator/Model/Formula.swift:
```diff
@@ -0,0 +1,40 @@
+			return lhsOperand / rhsOperand
+		case .multiply:
+			return lhsOperand * rhsOperand
+		}
```
`operators`내부에 `calculate`메서드를 구현하셨음에도 사용하시지 않은 이유가 궁금합니다!

wonbi92 commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -5,10 +5,6 @@
-protocol CalculateItem: Equatable { }
-
 struct CalculatorItemQueue<T: CalculateItem> {
 	private var list: LinkedList<T>?
```
제가 이부분을 지금 봤네요..! 리스트가 옵셔널인 이유가 있을꺼요?

wonbi92 reviewed (changes requested):
@devKobe24 
안녕하세요 코비! 스탭2 진행하시느라 고생하셨습니다!
코드에 고민하신 흔적이 잘 보여서 좋았습니다! 👍
특히 테스트 객체들과 테스트양을 보고 놀랐네요! 👀

몇가지 참고하실만한 부분들을 코맨트 드렸으니 쭉 읽어보시고 궁금하신 점 있으시면 말씀해주세요~ 💪

### 아래는 질문주신 부분에 대한 제 생각입니다! 😁
---
> 이번 스탭에서 가장 힘들었던 부분이 UML만 보고 내부 구현 코드를 만드는 것이였습니다. 그러면서 느낀점이 'UML을 보고 내부 구현을 어떻게 파악하지?' 가 가장 궁굼하고 조언을 얻고 싶은 부분이였습니다.
- 처음 UML을 보고 많이 당황하셨을거 같아요! 답답하신 마음 이해가 됩니다..! 🥹
- UML을 보시면서 내부 구현을 파악하려고 하시기 보다는 전체 앱의 구조들과 각 객체의 역할을 고민하시면 조금 도움이 되실거에요!
- 각각의 객체가 하는 일이 무엇이고, 서로 어떻게 의존하고 있는가를 파악해보세요! 그 부분이 파악되면 정말 자연스럽게 내부 구현이 어떻게 되어야 할지 가닥이 잡힐 거에요!

### 아래는 고민하신 부분에 대한 제 의견입니다. 😉
---
> 원비께서 주신 피드백 중 아에 다른 Mock 객체가 만들어졌다는 피드백을 듣고 실제 객체와 어떻게 연결해서 실험을 해야하는지 고민이 많았습니다. 제가 선택한 방법은 실제 객체 중 LinkedList 프로토콜을 Mock 객체와 연결하여 해결하였습니다.
- 테스트 더블 객체를 만드느라 많이 고민하신 흔적이 보여서 좋았습니다.
- 나름대로 프로토콜을 활용하고, 제네릭과 연관 타입을 처리하시려고 노력하신 부분 잘 보았습니다. 정말 고생 많으셨어요! 👏👏👏
- 사실 제가 말씀드렸던 부분은 `프로토콜로 추상화된 객체를 테스트 하는 것`입니다. 테스트는 실제 앱의 로직이 원하는대로 작동하는지 검증하는 것이죠?
- 한가지 간단한 예를 들어드리겠습니다.


1. 테스트하려는 객체를 `Queue`라고 해볼께요! 대표 로직으론 `enqueue,` `dequeue`등이 있겠네요.
2. 그러면 이 객체 자체를 인스턴스화 해서 테스트 할 수도 있겠지만, 제가 Mock을 만들어보라고 말씀 드렸죠?
3. Test Double을 제대로 진행하려면, `Queue`라는 객체를 추상화 할 필요가 있습니다. 이를 프로토콜로 추상화 해서 `Queueable`이라는 프로토콜을 만들었습니다. 당연히 `Queue`는 `Queueable`을 채택합니다.
```Swift
protocol Queueable {
    func enqueue(_ data: Int)
} 

struct Queue: Queueable {
    ...
}
```
4. 간단히 표현해봤습니다. 여기서 추상화 할 때 꼭 실제 앱에서 사용하는 로직들을 모두 명시해야 합니다.
5. 그럼 테스트할 때 테스트 할 객체를 만들어야 겠죠? `MockQueue`를 만들고 이 객체도 `Queueable`을 채택해야 합니다.
```Swift
struct MockQueue: Queueable {
    ...
}
```
6. 그 다음, 테스트 클래스에서 `sut`의 타입을 선언해주죠? 이때 `프로토콜을 타입으로 사용`합니다. 이 부분이 제가 말씀드리고자 했던 부분입니다. 이렇게하고 이를 사용할 때 우리가 만든 Mock 객체를 `주입`합니다.
```Swift
    var sut: (any Queueable)!

    override func setUpWithError() throws {
        try super.setUpWithError()
        sut = MockQueue()
    }
```
7. sut로 테스트를 진행하면 됩니다.

- 다음과 같은 순서로 진행하고 실제 앱에서도 Queue를 사용할 때 Queueable로 사용하고 실제 만든 객체를 주입해주면, 테스트 객체와 실제 앱에서 사용하는 객체의 의존도를 낮춰 결합도를 낮춰줍니다.
- 이게 SOLID원칙의 마지막 DIP, 즉 의존성 역전의 원칙입니다. 

---
- 정말 열심히 잘 해주셨어요. 다만, 로직이 제대로 돌아가지 않는 부분이 있어서 해당 부분은 수정이 조금 필요해보입니다. 마지막까지 화이팅입니다!! 💪

Commit:
:recycle: Refactor: Operator Extension 내부 코드 수정.

불필요한 operatorsCalculateResult 변수 삭제 및 코드 수정.

Commit:
:fire: Remove: import Foundation 삭제.

불필요한 import Foundation 삭제.

Commit:
:recycle: Refactor: 불필요한 상수 splitByTarget 삭제 및 코드 수정.

불필요한 상수 splitByTarget을 삭제하고 바로 return 값으로 self.split(separator: target).map { String($0) }을 주었습니다.

Commit:
:recycle: Refactor: result 내부 switch operators의 case 로직 변경

result 함수 내부에 switch operators의 case 로직을 operators.calculate(lhs: lhsOperand, rhs: rhsOperand) 를 각 case별로 return하도록 수정하였습니다.

Commit:
:sparkles: Feat: Queue를 만들기 위한 프로토콜 생성.

Queue를 추상화한 Queueable 프로토콜을 생성하였습니다.

Commit:
:recycle: Refactor: Queueable 프로토콜 채택 및 준수.

Queueable 프로토콜을 채택후 준수하였습니다.

Commit:
:art: Refactor: Queueable 프로토콜 생성, 채택, 준수.

Queable 프로토콜을 생성 후 CalculatorItemQueue에서 채택, 준수하였습니다.

Commit:
:recycle: Refactor: Queueable 채택 및 준수, 내부 코드 수정.

Queueable 프로토콜을 채택하고 준수하면서 생긴 오류 코드들에 대한 수정이 이루어졌습니다.

Commit:
:recycle: Refactor: 타입에 맞지 않는 코드 수정.

MockCalculatorItemQueue가 Queueable을 준수하게 되면서 타입이 맞지 않게된 타입들을 수정하였으며 init의 parameter name을 변경하였습니다.

Commit:
:recycle: Refactor: 테스트 케이스 전면 수정.

테스트 케이스에서 오류가 나는 부분 전면 수정하였습니다.
